IllegalStateException in akka.dispatch.BatchingExecutor$Batch.run

Our application provides REST interface to database storage. User uploads data via HTTP interface, we store it to JDBC and then do some additional logic that processes uploaded data. HTTP interface is implemented via Spray 1.3.2; Akka version 2.3.6 (but I have seen this problem in previous 2.3.x). We do not use akka-streams.

We create and then stop an some number of actors per upload; also we use transformation operations on Future in context dispatcher.

Sometimes we see this exception in our logs:

```
[incident-daemon-system-akka.actor.default-dispatcher-233] ERROR akka.dispatch.Dispatcher - exception in sameThreadExecutionContext (a.d.Dispatcher)
java.lang.IllegalStateException: exception in sameThreadExecutionContext
        at akka.dispatch.ExecutionContexts$sameThreadExecutionContext$.reportFailure(Future.scala:76)
        at scala.concurrent.impl.CallbackRunnable.executeWithValue(Promise.scala:40)
        at scala.concurrent.impl.Promise$DefaultPromise.tryComplete(Promise.scala:248)
        at scala.concurrent.Promise$class.complete(Promise.scala:55)
        at scala.concurrent.impl.Promise$DefaultPromise.complete(Promise.scala:153)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.Future$$anonfun$map$1.apply(Future.scala:235)
        at scala.concurrent.impl.CallbackRunnable.run(Promise.scala:32)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.processBatch$1(BatchingExecutor.scala:67)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:82)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:59)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:59)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
        at akka.dispatch.BatchingExecutor$Batch.run(BatchingExecutor.scala:58)
        at akka.dispatch.TaskInvocation.run(AbstractDispatcher.scala:41)
        at akka.dispatch.ForkJoinExecutorConfigurator$AkkaForkJoinTask.exec(AbstractDispatcher.scala:393)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
Caused by: java.lang.IllegalArgumentException: requirement failed
        at scala.Predef$.require(Predef.scala:207)
        at akka.dispatch.BatchingExecutor$Batch.run(BatchingExecutor.scala:55)
        at akka.dispatch.ExecutionContexts$sameThreadExecutionContext$.unbatchedExecute(Future.scala:74)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.processBatch$1(BatchingExecutor.scala:76)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply$mcV$sp(BatchingExecutor.scala:82)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:59)
        at akka.dispatch.BatchingExecutor$Batch$$anonfun$run$1.apply(BatchingExecutor.scala:59)
        at scala.concurrent.BlockContext$.withBlockContext(BlockContext.scala:72)
        at akka.dispatch.BatchingExecutor$Batch.run(BatchingExecutor.scala:58)
        at akka.dispatch.ExecutionContexts$sameThreadExecutionContext$.unbatchedExecute(Future.scala:74)
        at akka.dispatch.BatchingExecutor$class.execute(BatchingExecutor.scala:110)
        at akka.dispatch.ExecutionContexts$sameThreadExecutionContext$.execute(Future.scala:73)
        ... 19 common frames omitted
```

We see this exception when client continuously uploads large number of objects (about a million). I do not know the way to reproduce this bug (other that posting a lot of data and waiting before the problem occurs).

--- a/akka-actor/src/main/scala/akka/actor/ActorSystem.scala
+++ b/akka-actor/src/main/scala/akka/actor/ActorSystem.scala
@@ -602,6 +602,7 @@ private[akka] class ActorSystemImpl(val name: String, applicationConfig: Config,
     dynamicAccess.getObjectFor[ExecutionContext]("scala.concurrent.Future$InternalCallbackExecutor$").getOrElse(
       new ExecutionContext with BatchingExecutor {
         override protected def unbatchedExecute(r: Runnable): Unit = r.run()
+        override protected def resubmitOnBlock: Boolean = false // Since we execute inline, no gain in resubmitting
         override def reportFailure(t: Throwable): Unit = dispatcher reportFailure t
       })
 

--- a/akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala
+++ b/akka-actor/src/main/scala/akka/dispatch/AbstractDispatcher.scala
@@ -137,7 +137,7 @@ abstract class MessageDispatcher(val configurator: MessageDispatcherConfigurator
    * Detaches the specified actor instance from this dispatcher
    */
   final def detach(actor: ActorCell): Unit = try unregister(actor) finally ifSensibleToDoSoThenScheduleShutdown()
-
+  final protected def resubmitOnBlock: Boolean = true // We want to avoid starvation
   final override protected def unbatchedExecute(r: Runnable): Unit = {
     val invocation = TaskInvocation(eventStream, r, taskCleanup)
     addInhabitants(+1)

--- a/akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala
+++ b/akka-actor/src/main/scala/akka/dispatch/BatchingExecutor.scala
@@ -5,6 +5,7 @@
 package akka.dispatch
 
 import java.util.concurrent.{ Executor }
+import java.util.ArrayDeque
 import scala.concurrent._
 import scala.annotation.tailrec
 
@@ -45,41 +46,51 @@ private[akka] trait Batchable extends Runnable {
  */
 private[akka] trait BatchingExecutor extends Executor {
 
-  // invariant: if "_tasksLocal.get ne null" then we are inside BatchingRunnable.run; if it is null, we are outside
-  private val _tasksLocal = new ThreadLocal[List[Runnable]]()
+  // invariant: if "_tasksLocal.get ne null" then we are inside Batch.run; if it is null, we are outside
+  private[this] val _tasksLocal = new ThreadLocal[AbstractBatch]()
 
-  private class Batch(val initial: List[Runnable]) extends Runnable with BlockContext {
+  private[this] abstract class AbstractBatch extends ArrayDeque[Runnable](4) with Runnable {
+    @tailrec final def processBatch(batch: AbstractBatch): Unit =
+      if ((batch eq this) && !batch.isEmpty) {
+        batch.poll().run()
+        processBatch(_tasksLocal.get) // If this is null, then we have been using managed blocking, so bail out
+      }
+
+    protected final def resubmitUnbatched(): Boolean = {
+      val current = _tasksLocal.get()
+      _tasksLocal.remove()
+      if ((current eq this) && !current.isEmpty) { // Resubmit outselves if something bad happened and we still have work to do
+        unbatchedExecute(current) //TODO what if this submission fails?
+        true
+      } else false
+    }
+  }
+
+  private[this] final class Batch extends AbstractBatch {
+    override final def run: Unit = {
+      require(_tasksLocal.get eq null)
+      _tasksLocal set this // Install ourselves as the current batch
+      try processBatch(this) catch {
+        case t: Throwable ⇒
+          resubmitUnbatched()
+          throw t
+      } finally _tasksLocal.remove()
+    }
+  }
+
+  private[this] final class BlockableBatch extends AbstractBatch with BlockContext {
     private var parentBlockContext: BlockContext = _
     // this method runs in the delegate ExecutionContext's thread
-    override def run(): Unit = {
+    override final def run(): Unit = {
       require(_tasksLocal.get eq null)
-
+      _tasksLocal set this // Install ourselves as the current batch
       val prevBlockContext = BlockContext.current
       BlockContext.withBlockContext(this) {
-        try {
-          parentBlockContext = prevBlockContext
-
-          @tailrec def processBatch(batch: List[Runnable]): Unit = batch match {
-            case Nil ⇒ ()
-            case head :: tail ⇒
-              _tasksLocal set tail
-              try {
-                head.run()
-              } catch {
-                case t: Throwable ⇒
-                  // if one task throws, move the
-                  // remaining tasks to another thread
-                  // so we can throw the exception
-                  // up to the invoking executor
-                  val remaining = _tasksLocal.get
-                  _tasksLocal set Nil
-                  unbatchedExecute(new Batch(remaining)) //TODO what if this submission fails?
-                  throw t // rethrow
-              }
-              processBatch(_tasksLocal.get) // since head.run() can add entries, always do _tasksLocal.get here
-          }
-
-          processBatch(initial)
+        parentBlockContext = prevBlockContext
+        try processBatch(this) catch {
+          case t: Throwable ⇒
+            resubmitUnbatched()
+            throw t
         } finally {
           _tasksLocal.remove()
           parentBlockContext = null
@@ -89,13 +100,7 @@ private[akka] trait BatchingExecutor extends Executor {
 
     override def blockOn[T](thunk: ⇒ T)(implicit permission: CanAwait): T = {
       // if we know there will be blocking, we don't want to keep tasks queued up because it could deadlock.
-      {
-        val tasks = _tasksLocal.get
-        _tasksLocal set Nil
-        if ((tasks ne null) && tasks.nonEmpty)
-          unbatchedExecute(new Batch(tasks))
-      }
-
+      resubmitUnbatched()
       // now delegate the blocking to the previous BC
       require(parentBlockContext ne null)
       parentBlockContext.blockOn(thunk)
@@ -104,11 +109,16 @@ private[akka] trait BatchingExecutor extends Executor {
 
   protected def unbatchedExecute(r: Runnable): Unit
 
+  protected def resubmitOnBlock: Boolean
+
   override def execute(runnable: Runnable): Unit = {
     if (batchable(runnable)) { // If we can batch the runnable
       _tasksLocal.get match {
-        case null ⇒ unbatchedExecute(new Batch(List(runnable))) // If we aren't in batching mode yet, enqueue batch
-        case some ⇒ _tasksLocal.set(runnable :: some) // If we are already in batching mode, add to batch
+        case null ⇒
+          val newBatch: AbstractBatch = if (resubmitOnBlock) new BlockableBatch() else new Batch()
+          newBatch.add(runnable)
+          unbatchedExecute(newBatch) // If we aren't in batching mode yet, enqueue batch
+        case batch ⇒ batch.add(runnable) // If we are already in batching mode, add to batch
       }
     } else unbatchedExecute(runnable) // If not batchable, just delegate to underlying
   }

--- a/akka-actor/src/main/scala/akka/dispatch/Future.scala
+++ b/akka-actor/src/main/scala/akka/dispatch/Future.scala
@@ -72,6 +72,7 @@ object ExecutionContexts {
    */
   private[akka] object sameThreadExecutionContext extends ExecutionContext with BatchingExecutor {
     override protected def unbatchedExecute(runnable: Runnable): Unit = runnable.run()
+    override protected def resubmitOnBlock: Boolean = false // No point since we execute on same thread
     override def reportFailure(t: Throwable): Unit =
       throw new IllegalStateException("exception in sameThreadExecutionContext", t)
   }

--- a/project/AkkaBuild.scala
+++ b/project/AkkaBuild.scala
@@ -1084,8 +1084,16 @@ object AkkaBuild extends Build {
       
       // Change to improve cluster heartbeat sender, #16638
       FilterAnyProblem("akka.cluster.HeartbeatNodeRing"),
-      FilterAnyProblem("akka.cluster.ClusterHeartbeatSenderState")
-      
+      FilterAnyProblem("akka.cluster.ClusterHeartbeatSenderState"),
+
+      //Changes to improve BatchingExecutor, bugfix #16327
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor.resubmitOnBlock"),
+      ProblemFilters.exclude[FinalClassProblem]("akka.dispatch.BatchingExecutor$Batch"),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor#Batch.initial"),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor#Batch.blockOn"),
+      ProblemFilters.exclude[FinalMethodProblem]("akka.dispatch.BatchingExecutor#Batch.run"),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor#Batch.akka$dispatch$BatchingExecutor$Batch$$parentBlockContext_="),
+      ProblemFilters.exclude[MissingMethodProblem]("akka.dispatch.BatchingExecutor#Batch.this")
     )
   }
 

